### ISSUE TITLE
broken MIR in DropGlue(...) when compiling substrate (nightly)

### ISSUE THREAD

**Issue by naps62** [rust-lang/rust#105772]
### Code

Failed to compile substrate while following [the quick start guide](https://docs.substrate.io/quick-start/).
Compilation failed on rust nightly 1.67, worked on stable 1.66

```bash
git clone https://github.com/substrate-developer-hub/substrate-node-template
cd substrate-node-template
git checkout 10e62ab # main branch ref when this was reported
cargo build --package node-template --release
```

### Meta

`rustc --version --verbose`:
```
rustc 1.67.0-nightly (e0098a5cc 2022-11-29)
binary: rustc
commit-hash: e0098a5cc3a87d857e597af824d0ce1ed1ad85e0
commit-date: 2022-11-29
host: x86_64-unknown-linux-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

### Error output

```
   Compiling node-template v4.0.0-dev (/home/naps62/lightshift/odsy/substrate-node-template/node)
error: internal compiler error: no errors encountered even though `delay_span_bug` issued

error: internal compiler error: broken MIR in DropGlue(DefId(2:2801 ~ core[534f]::ptr::drop_in_place), Some(hyper::proto::h1::dispatch::Server<hyper::service::util::ServiceFn<fn(http::request::Request<jsonrpsee::jsonrpsee_server::logger::Body>) -> impl std::future::Future<Output = sc_servic
e::Result<http::response::Response<jsonrpsee::jsonrpsee_server::logger::Body>, std::convert::Infallible>> {jsonrpsee::jsonrpsee_server::transport::http::reject_connection::{closure#0}::reject}, jsonrpsee::jsonrpsee_server::logger::Body>, jsonrpsee::jsonrpsee_server::logger::Body>)) (after p
hase change to runtime-optimized) at bb0[0]:
                                Field projection `(*_1).field[0]` specified type `std::pin::Pin<std::boxed::Box<std::option::Option<[async fn body@jsonrpsee::jsonrpsee_server::transport::http::reject_connection::{closure#0}::reject::{closure#0}]>>>`, but actual type is `std::pin::Pin<std::b
oxed::Box<std::option::Option<<hyper::service::util::ServiceFn<fn(http::request::Request<jsonrpsee::jsonrpsee_server::logger::Body>) -> impl std::future::Future<Output = sc_service::Result<http::response::Response<jsonrpsee::jsonrpsee_server::logger::Body>, std::convert::Infallible>> {jsonr
psee::jsonrpsee_server::transport::http::reject_connection::{closure#0}::reject}, jsonrpsee::jsonrpsee_server::logger::Body> as hyper::service::http::HttpService<jsonrpsee::jsonrpsee_server::logger::Body>>::Future>>>`
   --> /home/naps62/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/lib/rustlib/src/rust/library/core/src/ptr/mod.rs:490:1
    |
490 | pub unsafe fn drop_in_place<T: ?Sized>(to_drop: *mut T) {
    | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: delayed at compiler/rustc_const_eval/src/transform/validate.rs:229:30

thread 'rustc' panicked at 'Box<dyn Any>', compiler/rustc_errors/src/lib.rs:1609:13


note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md

note: rustc 1.67.0-nightly (e0098a5cc 2022-11-29) running on x86_64-unknown-linux-gnu

note: compiler flags: --crate-type lib -C opt-level=3 -C embed-bitcode=no

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `node-template`
```

<details><summary><strong>Backtrace</strong></summary>
<p>

```
stack backtrace:
   0:     0x7f5155da25da - std::backtrace_rs::backtrace::libunwind::trace::h47611fed84c2afe1
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/../../backtrace/src/backtrace/libunwind.rs:93:5
   1:     0x7f5155da25da - std::backtrace_rs::backtrace::trace_unsynchronized::h196771ff1aeba450
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/../../backtrace/src/backtrace/mod.rs:66:5
   2:     0x7f5155da25da - std::sys_common::backtrace::_print_fmt::h0fbf9e8da54148fe
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/sys_common/backtrace.rs:65:5
   3:     0x7f5155da25da - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h2d625fe89dc9be85
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/sys_common/backtrace.rs:44:22
   4:     0x7f5151d4612e - core::fmt::write::ha2a0d746f0011cf0
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/core/src/fmt/mod.rs:1208:17
   5:     0x7f5155d967a5 - std::io::Write::write_fmt::h4d18be5583014ce7
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/io/mod.rs:1682:15
   6:     0x7f5155da23a5 - std::sys_common::backtrace::_print::h3f824f2a08775117
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/sys_common/backtrace.rs:47:5
   7:     0x7f5155da23a5 - std::sys_common::backtrace::print::h3a59e28067a1eeb3
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/sys_common/backtrace.rs:34:9
   8:     0x7f5155da470f - std::panicking::default_hook::{{closure}}::h75a2d2e7583d6c9d
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/panicking.rs:267:22
   9:     0x7f5155da444b - std::panicking::default_hook::hbd78103b5e047f9f
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/panicking.rs:286:9
  10:     0x7f5154fedbe4 - <rustc_driver[bb1da67439e462ba]::DEFAULT_HOOK::{closure#0}::{closure#0} as core[534fe282bae85850]::ops::function::FnOnce<(&core[534fe282bae85850]::panic::panic_info::PanicInfo,)>>::call_once::{shim:vtable#0}
  11:     0x7f5155da4f0d - <alloc::boxed::Box<F,A> as core::ops::function::Fn<Args>>::call::he770e6f3ac0a3b2a
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/alloc/src/boxed.rs:2032:9
  12:     0x7f5155da4f0d - std::panicking::rust_panic_with_hook::ha807f3b0ca9de4b9
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/panicking.rs:692:13
  13:     0x7f51550265a1 - std[7cf7edf1b9174793]::panicking::begin_panic::<rustc_errors[556381c31637ba53]::ExplicitBug>::{closure#0}
  14:     0x7f5155024e66 - std[7cf7edf1b9174793]::sys_common::backtrace::__rust_end_short_backtrace::<std[7cf7edf1b9174793]::panicking::begin_panic<rustc_errors[556381c31637ba53]::ExplicitBug>::{closure#0}, !>
  15:     0x7f515501da26 - std[7cf7edf1b9174793]::panicking::begin_panic::<rustc_errors[556381c31637ba53]::ExplicitBug>
  16:     0x7f5155021296 - std[7cf7edf1b9174793]::panic::panic_any::<rustc_errors[556381c31637ba53]::ExplicitBug>
  17:     0x7f515450d8e6 - <rustc_errors[556381c31637ba53]::HandlerInner>::flush_delayed::<alloc[d2ee30a43ea25fa3]::vec::Vec<rustc_errors[556381c31637ba53]::diagnostic::Diagnostic>, &str>
  18:     0x7f515450cbeb - <rustc_errors[556381c31637ba53]::HandlerInner as core[534fe282bae85850]::ops::drop::Drop>::drop
  19:     0x7f515425a76e - core[534fe282bae85850]::ptr::drop_in_place::<rustc_session[eb3cf40b531044b2]::parse::ParseSess>
  20:     0x7f515424b42f - core[534fe282bae85850]::ptr::drop_in_place::<rustc_session[eb3cf40b531044b2]::session::Session>
  21:     0x7f515424b160 - core[534fe282bae85850]::ptr::drop_in_place::<rustc_interface[d6dbbac1f1e880d6]::interface::Compiler>
  22:     0x7f515424a26e - rustc_span[c61ccd2374fbbf2a]::with_source_map::<core[534fe282bae85850]::result::Result<(), rustc_errors[556381c31637ba53]::ErrorGuaranteed>, rustc_interface[d6dbbac1f1e880d6]::interface::run_compiler<core[534fe282bae85850]::result::Result<(), rustc_errors[556381c3
1637ba53]::ErrorGuaranteed>, rustc_driver[bb1da67439e462ba]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  23:     0x7f5154249ba5 - <scoped_tls[3b1485898944538b]::ScopedKey<rustc_span[c61ccd2374fbbf2a]::SessionGlobals>>::set::<rustc_interface[d6dbbac1f1e880d6]::interface::run_compiler<core[534fe282bae85850]::result::Result<(), rustc_errors[556381c31637ba53]::ErrorGuaranteed>, rustc_driver[bb1d
a67439e462ba]::run_compiler::{closure#1}>::{closure#0}, core[534fe282bae85850]::result::Result<(), rustc_errors[556381c31637ba53]::ErrorGuaranteed>>
  24:     0x7f5154249192 - std[7cf7edf1b9174793]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[d6dbbac1f1e880d6]::util::run_in_thread_pool_with_globals<rustc_interface[d6dbbac1f1e880d6]::interface::run_compiler<core[534fe282bae85850]::result::Result<(), rustc_errors
[556381c31637ba53]::ErrorGuaranteed>, rustc_driver[bb1da67439e462ba]::run_compiler::{closure#1}>::{closure#0}, core[534fe282bae85850]::result::Result<(), rustc_errors[556381c31637ba53]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[534fe282bae85850]::result::Result<(), rustc_errors[5563
81c31637ba53]::ErrorGuaranteed>>
  25:     0x7f5154248ea8 - <<std[7cf7edf1b9174793]::thread::Builder>::spawn_unchecked_<rustc_interface[d6dbbac1f1e880d6]::util::run_in_thread_pool_with_globals<rustc_interface[d6dbbac1f1e880d6]::interface::run_compiler<core[534fe282bae85850]::result::Result<(), rustc_errors[556381c31637ba53
]::ErrorGuaranteed>, rustc_driver[bb1da67439e462ba]::run_compiler::{closure#1}>::{closure#0}, core[534fe282bae85850]::result::Result<(), rustc_errors[556381c31637ba53]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[534fe282bae85850]::result::Result<(), rustc_errors[556381c31637ba53]::Er
rorGuaranteed>>::{closure#1} as core[534fe282bae85850]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  26:     0x7f5155dac143 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h4589c8a57cba1bcb
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/alloc/src/boxed.rs:2000:9
  27:     0x7f5155dac143 - <alloc::boxed::Box<F,A> as core::ops::function::FnOnce<Args>>::call_once::h11590d2327f9fc1e
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/alloc/src/boxed.rs:2000:9
  28:     0x7f5155dac143 - std::sys::unix::thread::Thread::new::thread_start::hd9e44ab9bbc3123f
                               at /rustc/e0098a5cc3a87d857e597af824d0ce1ed1ad85e0/library/std/src/sys/unix/thread.rs:108:17
  29:     0x7f5151a29b43 - start_thread
                               at ./nptl/./nptl/pthread_create.c:442:8
  30:     0x7f5151abba00 - clone3
                               at ./misc/../sysdeps/unix/sysv/linux/x86_64/clone3.S:81
  31:                0x0 - <unknown>
```

</p>
</details>

**Comment by matthiaskrgr** [comment]
>2022-11-29
your nightly is half a month old. Could you try with a recent one?

**Comment by naps62** [comment]
uh. I could have swore I updated before. guess I messed up something, my bad
